nexus/db-queries/src/db/datastore/dataset.rs

```diff
@@ -73,9 +82,261 @@ impl DataStore {
                 e,
                 ErrorHandler::Conflict(
                     ResourceType::Dataset,
-                    &dataset.id().to_string(),
+                    &dataset_id.to_string(),
                 ),
             ),
         })
     }
+
+    /// Stores a new dataset in the database, but only if a dataset with the
+    /// given `id` does not already exist
+    ///
+    /// Does not update existing rows. If a dataset with the given ID already
+    /// exists, returns `Ok(None)`.
+    pub async fn dataset_insert_if_not_exists(
```

Comment:
This seems like it'll fit in well with the existing "sled agent inserts dataset" structure, as we transition more towards "Nexus controls dataset creation".